include/zephyr/drivers/battery.h

```diff
@@ -0,0 +1,210 @@
+/* /\** General Error/Runtime Flags *\/ */
+/* #define BATTERY_FLAGS		     BATTERY_CONNECT_STATE + 1 */
+/* /\** Full Charge Capacity in mAh (might change in some implementations) *\/ */
+/* #define BATTERY_FULL_CHARGE_CAPACITY BATTERY_FLAGS + 1 */
+/* /\** */
+/*  * Report the absolute difference between the highest and lowest cell voltage in */
+/*  * the battery pack, in millivolts.  On error returns '0'. */
+/*  *\/ */
+/* #define BATTERY_IMBALANCE_MV	     BATTERY_FULL_CHARGE_CAPACITY + 1 */
+/* /\** Is the battery physically present *\/ */
+/* #define BATTERY_PRESENT_STATE	     BATTERY_IMBALANCE_MV + 1 */
+/* /\** Remaining capacity in mAh *\/ */
+/* #define BATTERY_REMAINING_CAPACITY   BATTERY_PRESENT_STATE + 1 */
+/* /\** State of charge (percent, 0-100) *\/ */
+/* #define BATTERY_STATE_OF_CHARGE	     BATTERY_REMAINING_CAPACITY + 1 */
+/* /\** Temperature in 0.1 K *\/ */
+/* #define BATTERY_TEMPERATURE	     BATTERY_STATE_OF_CHARGE + 1 */
+/* /\** Battery voltage (mV) *\/ */
+/* #define BATTERY_VOLTAGE		     BATTERY_TEMPERATURE + 1 */
+
+/* /\** Reserved to demark downstream custom properties are being detched *\/ */
+/* #define BATTERY_CUSTOM_BEGIN = 256 */
+
+/* /\** END BATTERY PROPERTIES *\/ */
+
+struct battery_get_property {
+	/** Battery property to get */
+	struct {
+		bool use_cached : 1;
+		/** Set by callee */
+		bool is_valid : 1;
+		enum battery_proprty_type property_type : 30;
+	};
+
+	/** Property field for getting */
+	union {
+		/** Fields have the format: */
+		/** BATTERY_PROPERTY_FIELD */
+		/** type property_field; */
+
+		/** Dynamic Battery Info */
+		/** BATTERY_AVG_CURRENT, */
+		int avg_current;
+		/** BATTERY_AVG_VOLTAGE, */
+		int avg_voltage;
+		/** BATTERY_CURRENT, */
+		int current;
+		/** BATTERY_CUT_OFF, */
+		bool cut_off;
+		/** BATTERY_CYCLE_COUNT, */
+		uint32_t cycle_count;
+		/** BATTERY_CONNECT_STATE, */
+		enum battery_connect_state connect_state;
+		/** BATTERY_FLAGS, */
+		uint32_t flags;
+		/** BATTERY_FULL_CHARGE_CAPACITY, */
+		uint32_t full_charge_capacity;
+		/** BATTERY_IMBALANCE_MV, */
+		uint32_t imbalance_mv;
+		/** BATTERY_REMAINING_CAPACITY, */
+		uint32_t remaining_capacity;
+		/** BATTERY_RUN_TIME_TO_EMPTY, */
+		uint32_t run_time_to_empty;
+		/** BATTERY_STATE_OF_CHARGE, */
+		uint8_t state_of_charge;
+		/** BATTERY_TEMPERATURE, */
+		uint32_t temperature;
+		/** BATTERY_VOLTAGE, */
+		int voltage;
+	};
+};
+
+/**
+ * @brief Get a battery property
+ *
+ * @param dev Pointer to the battery device
+ * @param props pointer to array of battery_get_property struct where the property struct
+ * field is set by the caller to determine what property is read from the
+ * battery device.
+ * @param props_len number of properties in props array
+ *
+ * @return 0 if successful, negative errno code if failure.
+ */
+typedef int (*battery_get_property_t)(const struct device *dev, struct battery_get_property *props,
+				      size_t props_len);
+
+/* TODO(aaronmassey) Probably remove */
```

Comment:
+1 to that, -ENOTSUP from get_property is fine for indicating this and makes the API even simpler